### PR TITLE
Add docs about DB queries scoping with Pop

### LINKED ISSUE #181: Documentation / tutorial Scoping
The Scoping function in Buffalo needs a documentation.
----------------------------------------------------------------------------
Here is a draft / reminder (for me ):

Once you have been through the tutorial on how to set a login on your app using Goth or Password auth (
https://gobuffalo.io/docs/examples#using-goth-with-buffalo
https://gobuffalo.io/docs/examples#using-password-authentication-with-buffalo
)
chances are that you want to limit the access of certain resources to certain users.
For example you don t want everyone to be able to modify the profile of every one on your website!
You want each user to be able to change ONLY the parameter of its own profile.
This is done using scoping.
[to be continued ...]

-------------

Stanislas technical  def  : 

The scope function returns a pop query builder. Its aim is to build queries with the same base.
So `tx := c.Value("tx").(*pop.Connection)` gets the current pop transaction.
Then `cu, ok := c.Value("current_user").(*models.User)` gets the current user from the context.
You can get it from there, because a middleware set it from  the session attribute `current_user_id`.
For reference, this middleware is declared here https://github.com/gobuffalo/toodo/blob/master/actions/users.go#L43.
The `cu, ok := c.Value("current_user").(*models.User)` returns two values: one to get the current user if it was set (= the user is connected), and `ok` is set to `true` if the current user is a valid user.
If it's not `ok`, the scope function returns an empty query constructor using `tx.Q()`.
It means there is no current user, so you can't filter it.

Otherwise, it'll return a new query constructor with a filter preset: `tx.BelongsTo(cu)` will create a base query with a WHERE user_id = current user id.
So for each query using this preset constructor (aka scope), it will add a WHERE user_id = current user id if there is a connected user.
Or default to the non-filtered query otherwise.

### PR DESCRIPTION
**Pull request by stanislas-m**
Fixes #181.

### CODE REVIEW

File: templates/docs/db/scoping.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    fmt.Printf("%v\n", book)
+}
+```
+
+<%= title("The scoped way") %>
+
+The scope factorizes the common part of the query:
+
+```go
+type struct Book {
```
Comment (markbates) [inline templates/docs/db/scoping.md]: `type Book struct`

File: templates/docs/db/scoping.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
+# Scoping
+
+Scoping is a way to structure your DB calls, when it needs the same "base" query. Let's say you want to create a book store: the store provides books for everyone, but some special editions are reserved to customers with a registered account. It means that for the whole store, you'll need to filter the books, so the "guest" customers can only see the restricted list of books.
+
+<%= title("The usual way") %>
+
+A "naive" way can be writing each full query.
+
+```go
+type struct Book {
```
Comment (markbates) [inline templates/docs/db/scoping.md]: `type Book struct`

File: templates/docs/db/scoping.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    ID         uuid.UUID `json:"id" db:"id"`
+    Label      string    `json:"label" db:"label"`
+    Restricted bool      `json:"is_restricted" db:"is_restricted"`
+}
+
+type Books []Book
+```
+
+```go
+func restrictedScope(c buffalo.Context, registeredAccount bool) *pop.Query {
```
Comment (markbates) [inline templates/docs/db/scoping.md]: you may want to remove the buffalo dependency here.

I would probably rewrite this to be a "pure" `pop.ScopeFunc`

```go
func restrictedScope(registeredAccount bool) pop.ScopeFunc {
  return func(q *pop.Query) *pop.Query {
    if !registeredAccount {
      return q
    }
    return q.Where("is_restricted = false")
  }
}

if err := tx.Scope(restrictedScope(registeredAccount)).All(&books); err != nil {
}
```

Comment (stanislas-m) [reply templates/docs/db/scoping.md]: Nice! I didn't know about this one. :+1:

**Pull request merged**